cache transform when computing geometries, probably more low hanging …

erh opened the pull request:
…fruit

erh committed:
cache transform when computing geometries, probably more low hanging fruit

erh committed:
cache more

erh committed:
lint and test fix

dgottlieb reviewed (approved):
Didn't check the callsite updates to check for cache invalidation -- just looked at the composeTransforms bit. Let me know if you think there's value in me taking a closer look for bugs in the usage. I assumed that class of bugs wasn't particularly at risk here.

erh committed:
comment

Update comment for composeTransforms function to clarify its purpose and cache behavior.

erh commented:
> Didn't check the callsite updates to check for cache invalidation -- just looked at the composeTransforms bit. Let me know if you think there's value in me taking a closer look for bugs in the usage. I assumed that class of bugs wasn't particularly at risk here.

this had less impact than i expected.
can you take a closer look at some point and see what you think.

erh committed:
Merge branch 'viamrobotics:main' into 20251107-cache-transforms

erh committed:
Merge remote-tracking branch 'origin/main' into 20251107-cache-transforms

erh committed:
fix test, still missing things

dgottlieb commented:
> can you take a closer look at some point and see what you think.

Totally lost track of this. 
TL;DR:
* I think this PR was attacking calls to transforming geometries
  * I found instrumented caches in [metrics.go](https://github.com/viamrobotics/rdk/pull/5456/changes#diff-1947092b7a3d7686347352842406b6926069272d2dabfe8af425b36b38da5960R71) (read: constraint checker)
  * and (simple)model.go that manifests as this [loop-cache](https://github.com/viamrobotics/rdk/pull/5456/changes#diff-e09745754de51ce050479bd82d1104757e21f370d57cde2659135a045c13789fR680)
* For those callers -- I think the cost of caching is only a bit better than redoing the computation.
  * For example when I instead cache the entire dualquat.Number output for a `left-gripper`, I get a 10x speedup for a cache hit.
  * Meaning each frame cache hit in the original PR is only a <2x speedup.

I also found more calls to `composeTransforms` was coming from IK (getGoalMetrics result lambda from `planner_options.go`). So I tried to make that fast:
* I tuned down IK to a single CPU to get less noisy measurements
* goal metrics (read:  IK/nlopt) represent 90% of the `composeTransforms` in most scenarios
  * That was not instrumented by the PR
  * I hacked in a global cache mapping inputs -> cache
  * There are some gains to be had by instrumenting it -- but it's not as simple as using the input `StateFS` cache. If I interpreted things correctly, IK rounds often independently get the same solutions.
* Given the individual frame cache hit performance is small
  * I made a change to instead cache the final value that `composeTransforms` returns.
    * This resulted in cache hits being ~10x faster than cache misses (for getGoalMetrics)
    * Losing some, but not all, of intermediate results/benefits that are relevant to the other code paths
    * No reason we can't have both
* Many tests don't spend a lot of time in composeTransforms anyways (easier to see with nCPU=1)
  * TestWineCrazyTouch2 is an example of a test that spends (relatively) a lot of time in `composeTransforms`: ~10% (250ms per each of three subtest that each take ~1s)
  * With caching that gets us a nearly 70% hit rate, that cost goes down to ~100ms per subtest.
  * Resulting in a modest ~10% savings for this best case scenario.

Another thing I scrutinized (with my planner_options.go addition that also did not yield a big improvement) was that any time saved in an individual IK computation would just result in more time being spent doing more IK rounds. That didn't seem to hold.

The longer raw-form/with data:

Looked at this (a bit too much) the last day. Some observations:
* I also played with caching just the final result for the input frame.
  * These numbers are with that change -- broadly I found that the cache lookup for a frame -> its parent is cheaper, but still comparable, to the computation the cache is avoiding.
* I changed [nCPU](https://github.com/viamrobotics/rdk/blob/d343edabb8b9f84c01abb3697ed1a36981cb1d48/motionplan/ik/combined.go#L35) down to 1 to get more accurate "CPU" time.
  * Lots of happy path cases just need 1 solution to move forward. And it takes one IK round to get that one good solution. Having multiple CPUs "repeat" that work doesn't change the latency, but adds noise to how much time was spent in IK measurements.
* In that 1 CPU case, `TestWineCrazyTouch2` takes:
```
--- PASS: TestWineCrazyTouch2 (3.48s)
    --- PASS: TestWineCrazyTouch2/regular (1.05s)
    --- PASS: TestWineCrazyTouch2/regular-noclose (0.94s)
    --- PASS: TestWineCrazyTouch2/orientation (1.49s)
```
* `TestWineCrazyTouch2` is a bit interesting because we spend a non-trivial amount of time in `composeTransforms`
  * Without any caching the amount of `composeTransforms` time is `725.128656ms`
  * Doing IK for each of the three subtests:
```
2026-05-05T12:49:33.297-0400	INFO	TestWineCrazyTouch2.mp.solve	armplanning/plan_manager.go:323	DBG. NumSols: 7 TimeSpent: 702.40584ms
2026-05-05T12:49:34.318-0400	INFO	TestWineCrazyTouch2.mp.solve	armplanning/plan_manager.go:323	DBG. NumSols: 9 TimeSpent: 671.608249ms
2026-05-05T12:49:35.262-0400	INFO	TestWineCrazyTouch2.mp.solve	armplanning/plan_manager.go:323	DBG. NumSols: 9 TimeSpent: 676.375961ms
```
* The three main callers of `composeTransform`, with cache statistics, but we aren't using the cached values, as a baseline:
```
Caller: /home/dgottlieb/viam/rdk/motionplan/armplanning/planner_options.go:168 Stats:
	 Calls: 365550 Time: 541.948697ms PerCall: 1.482µs Nils: 0 UniqueCnt: 120833 Calls/Unique: 3 Hit%: 66.94% IsSeenMiss: 244717
Caller: /home/dgottlieb/viam/rdk/referenceframe/model.go:588 Stats:
	 Calls: 211662 Time: 104.630202ms PerCall: 494ns Nils: 0 UniqueCnt: 5863 Calls/Unique: 36 Hit%: 86.15% IsSeenMiss: 205799
Caller: /home/dgottlieb/viam/rdk/motionplan/metrics.go:72 Stats:
	 Calls: 115380 Time: 74.076674ms PerCall: 642ns Nils: 0 UniqueCnt: 5198 Calls/Unique: 22 Hit%: 45.94% IsSeenMiss: 110182
```
* When we turn on caching:
```
Caller: /home/dgottlieb/viam/rdk/motionplan/armplanning/planner_options.go:168 Stats:
	 Calls: 352997 Time: 287.725685ms PerCall: 815ns Nils: 0 UniqueCnt: 121702 Calls/Unique: 2 Hit%: 65.52% IsSeenMiss: 0
Caller: /home/dgottlieb/viam/rdk/referenceframe/model.go:588 Stats:
	 Calls: 211662 Time: 61.010295ms PerCall: 288ns Nils: 0 UniqueCnt: 5863 Calls/Unique: 36 Hit%: 86.15% IsSeenMiss: 23449
Caller: /home/dgottlieb/viam/rdk/motionplan/metrics.go:72 Stats:
	 Calls: 115380 Time: 57.846389ms PerCall: 501ns Nils: 0 UniqueCnt: 5198 Calls/Unique: 22 Hit%: 45.94% IsSeenMiss: 57178
```
* Note that `planner_options` was not instrumented in the PR. But getting this hit rate required a global inputs->transformed `dualquat.Number` cache, not the localized one passed around on [the available `stateFS`](https://github.com/viamrobotics/rdk/pull/5456/changes#diff-4660f93c6d15c86f7749692e5164d0e0a702f6a2a7657f82b366654603ddb3f4R164) object.
* And more aggregate stats:
```
	composeTransforms time -- CacheHit: 0s, CacheMiss: 725.128656ms
vs:
	composeTransforms time -- CacheHit: 107.68548ms, CacheMiss: 295.340114ms
```
* So while cache hits shave off ~330ms in total, the average time to compute is only halved.
* And this was one of the best cases for the patch as is (no planner_options.go caching)
  * crazy touch 2 has a ~3:2:1 ratio for planner options -> model.go -> metrics.go calls.
  * In most other tests > 90% of the `composeTransforms` calls from planner_options.go.
